File: app/src/main/java/com/aquamorph/frcmanager/Fragments/AwardFragment.java
package com.aquamorph.frcmanager.fragments;

import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.aquamorph.frcmanager.Divider;
import com.aquamorph.frcmanager.R;
import com.aquamorph.frcmanager.adapters.AwardAdapter;
import com.aquamorph.frcmanager.models.Award;
import com.aquamorph.frcmanager.parsers.AwardParser;

import java.util.ArrayList;

/**
 * Displays a list of awards at a event
 *
 * @author Christian Colglazier
 * @version 2/13/2016
 */
public class AwardFragment extends Fragment implements SharedPreferences.OnSharedPreferenceChangeListener {

	private String TAG = "AwardFragment";
	private SwipeRefreshLayout swipeRefreshLayout;
	private RecyclerView recyclerView;
	private TextView emptyView;
	private RecyclerView.Adapter adapter;
	private ArrayList<Award> awards = new ArrayList<>();
	private String eventKey = "";
	SharedPreferences prefs;

	/**
	 * newInstance creates and returns a new AwardFragment
	 *
	 * @return AwardFragment
	 */
	public static AwardFragment newInstance() {
		return new AwardFragment();
	}

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setRetainInstance(true);
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_team_schedule, container, false);
		swipeRefreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.swipeRefreshLayout);
		swipeRefreshLayout.setColorSchemeResources(R.color.accent);
		swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
			@Override
			public void onRefresh() {
				refresh();
			}
		});

		recyclerView = (RecyclerView) view.findViewById(R.id.rv);
		emptyView = (TextView) view.findViewById(R.id.empty_view);
		adapter = new AwardAdapter(getContext(), awards);
		LinearLayoutManager llm = new LinearLayoutManager(getContext());
		llm.setOrientation(LinearLayoutManager.VERTICAL);
		recyclerView.setAdapter(adapter);
		recyclerView.setLayoutManager(llm);
		recyclerView.addItemDecoration(new Divider(getContext()));

		prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
		prefs.registerOnSharedPreferenceChangeListener(AwardFragment.this);
		eventKey = prefs.getString("eventKey", "");

		refresh();
		return view;
	}

	/**
	 * refrest() loads data needed for this fragment.
	 */
	public void refresh() {
		if (!eventKey.equals("")) {
			final LoadAwards loadAwards = new LoadAwards();
			loadAwards.execute();
		}
	}

	@Override
	public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
		if (key.equals("eventKey")) {
			eventKey = sharedPreferences.getString("eventKey", "");
			refresh();
		}
	}

	class LoadAwards extends AsyncTask<Void, Void, Void> {

		@Override
		protected void onPreExecute() {
			swipeRefreshLayout.setRefreshing(true);
		}

		@Override
		protected Void doInBackground(Void... params) {
			AwardParser awardParser = new AwardParser();
			awardParser.fetchJSON(eventKey, getContext());
			while (awardParser.parsingComplete) ;
			awards.clear();
			awards.addAll(awardParser.getAwards());
			return null;
		}

		@Override
		protected void onPostExecute(Void result) {
			adapter.notifyDataSetChanged();
			if (awards.isEmpty()) {
				recyclerView.setVisibility(View.GONE);
				emptyView.setVisibility(View.VISIBLE);
			}
			else {
				recyclerView.setVisibility(View.VISIBLE);
				emptyView.setVisibility(View.GONE);
			}
			swipeRefreshLayout.setRefreshing(false);
		}
	}
}

File: app/src/main/kotlin/com/aquamorph/frcmanager/fragments/AwardFragment.kt
package com.aquamorph.frcmanager.fragments

import android.content.SharedPreferences
import android.os.Bundle
import android.os.SystemClock
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.aquamorph.frcmanager.R
import com.aquamorph.frcmanager.adapters.AwardAdapter
import com.aquamorph.frcmanager.decoration.Animations
import com.aquamorph.frcmanager.decoration.Divider
import com.aquamorph.frcmanager.models.tba.Award
import com.aquamorph.frcmanager.network.DataLoader
import com.aquamorph.frcmanager.utils.Constants

/**
 * Displays a list of awards at a event
 *
 * @author Christian Colglazier
 * @version 3/25/2023
 */
class AwardFragment :
        TabFragment(), SharedPreferences.OnSharedPreferenceChangeListener, RefreshFragment {

    private var awards: ArrayList<Award> = ArrayList()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_team_schedule, container, false)
        super.onCreateView(view, awards,
                AwardAdapter(requireContext(), awards),
                Divider(requireContext(), Constants.DIVIDER_WIDTH, 0))
        prefs.registerOnSharedPreferenceChangeListener(this)
        return view
    }

    override fun dataUpdate() {
        awards.clear()
        awards.addAll(DataLoader.awardDC.data)
        filterAwards()
        Constants.checkNoDataScreen(DataLoader.awardDC.data, recyclerView, emptyView)
        Animations.loadAnimation(context, recyclerView, adapter,
                firstLoad, DataLoader.awardDC.newData)
        firstLoad = false
    }

    /**
     * filterAwards() removes awards that were not given out.
     */
    private fun filterAwards() {
        var i = 0
        while (i < awards.size) {
            if (awards[i].recipientList.size == 1 &&
                    awards[i].recipientList[0].awardee == null &&
                    awards[i].recipientList[0].teamKey == null) {
                awards.remove(awards[i])
            } else {
                i++
            }
        }
    }

    override fun onResume() {
        super.onResume()
        if (awards.isEmpty())
            refresh()
    }

    /**
     * refresh() loads dataLoader needed for this fragment.
     */
    override fun refresh() {
        if (DataLoader.eventKey != "") {
            mSwipeRefreshLayout.isRefreshing = true
            executor.execute {
                while (!DataLoader.awardDC.complete) {
                    SystemClock.sleep(Constants.THREAD_WAIT_TIME.toLong())
                }
                handler.post {
                    if (context != null) {
                        dataUpdate()
                        mSwipeRefreshLayout.isRefreshing = false
                    }
                }
            }
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String?) {
        if (key == "eventKey") {
            refresh()
        }
    }

    companion object {

        /**
         * newInstance creates and returns a new AwardFragment
         *
         * @return AwardFragment
         */
        fun newInstance(): AwardFragment {
            return AwardFragment()
        }
    }
}
